File: windows/src/AINotebook.Core/Providers/ProviderType.cs
namespace AINotebook.Core.Providers;

public enum ProviderType { Ollama, Anthropic, OpenAI, OpenAICompatible, OpenWebUI }

public static class ProviderTypeExtensions
{
    public static string ToStorageString(this ProviderType t) => t switch
    {
        ProviderType.Ollama => "ollama",
        ProviderType.Anthropic => "anthropic",
        ProviderType.OpenAI => "openai",
        ProviderType.OpenAICompatible => "openai_compatible",
        ProviderType.OpenWebUI => "openwebui",
        _ => throw new ArgumentOutOfRangeException(nameof(t))
    };

    public static ProviderType FromStorageString(string s) => s switch
    {
        "ollama" => ProviderType.Ollama,
        "anthropic" => ProviderType.Anthropic,
        "openai" => ProviderType.OpenAI,
        "openai_compatible" => ProviderType.OpenAICompatible,
        "openwebui" => ProviderType.OpenWebUI,
        _ => ProviderType.OpenAICompatible
    };

    public static string DefaultBaseUrl(this ProviderType t) => t switch
    {
        ProviderType.Ollama => "http://127.0.0.1:11434",
        ProviderType.Anthropic => "https://api.anthropic.com",
        ProviderType.OpenAI => "https://api.openai.com",
        ProviderType.OpenAICompatible => "",
        ProviderType.OpenWebUI => "",
        _ => ""
    };

    // OpenWebUI has no OpenAI-compatible embeddings endpoint — chat only.
    public static bool SupportsEmbeddings(this ProviderType t) =>
        t != ProviderType.Anthropic && t != ProviderType.OpenWebUI;
}

File: Sources/AINotebookCore/Providers/ProviderType.swift
import Foundation

/// Provider registry types. Storage strings match the Windows port exactly
/// (`windows/src/AINotebook.Core/Providers/ProviderType.cs`).
public enum ProviderType: String, CaseIterable, Sendable {
    case ollama
    case anthropic
    case openai
    case openaiCompatible = "openai_compatible"
    case openwebui

    /// Unknown storage strings fall back to `.openaiCompatible` (Windows parity).
    public static func fromStorage(_ raw: String) -> ProviderType {
        ProviderType(rawValue: raw) ?? .openaiCompatible
    }

    public var defaultBaseURL: String {
        switch self {
        case .ollama: "http://127.0.0.1:11434"
        case .anthropic: "https://api.anthropic.com"
        case .openai: "https://api.openai.com"
        case .openaiCompatible, .openwebui: ""
        }
    }

    /// Anthropic has no embeddings API; OpenWebUI is chat-only by design.
    public var supportsEmbeddings: Bool {
        switch self {
        case .anthropic, .openwebui: false
        case .ollama, .openai, .openaiCompatible: true
        }
    }

    /// True when requests leave this machine — the privacy gate applies.
    public var isCloud: Bool { self != .ollama }
}
